### ISSUE TITLE
Add support for splats in shapes

### ISSUE THREAD

**Issue by matrixise** [edgedb/edgedb#180]
Is there a way to get all the attributes of an object, maybe on the first level.

these statements do not work, (not explained in the documentation).
```
SELECT User {};
SELECT User {*};
```

**Comment by elprans** [comment]
Currently, you have to specify the output shape explicitly. The main reason is that it is not obvious what is meant by `*`. Is it all properties, or all properties and links, or all singleton properties?  All three have different performance characteristics.

Generally speaking, `*` is mostly only useful in repl, where you need to run something quickly. We might think of something for repl specifically.

**Comment by elprans** [comment]
The proposed solution is to provide an autocomplete shortcut so that typing `SELECT User{*<tab>` would complete to a full list of type properties (but not links).

**Comment by vpetrovykh** [comment]
Just as a reminder, we also considered to use `...<tab>` for that autocomplete shortcut. The benefit of that is largely that we don't have to disambiguate whether or not the `*` appears syntactically as an operator or not.

**Comment by 1st1** [comment]
Most likely it will be `*<TAB>`

**Comment by 1st1** [comment]
We're now considering adding support for splats in shapes:

```
SELECT User { ...User };
# or
SELECT User { ...UserFragmentView };
```

Using `...` makes more sense than `*` since it's more readable to a non-Python programmer and is similar to the corresponding GraphQL syntax.

**Comment by vpetrovykh** [comment]
A few of nuances need to be addressed in regards to using views vs. types in splats.

## 1. What is included by a splat

The simple case:
```
# type is splatted, so all of its properties and links are included
SELECT User {...User};

# view is splatted, so only the explicitly defined shape elements are included?
WITH 
    UserView := User {
        name,
        foo := 'bar'
    }
SELECT User {...UserView};
```

## 2. Ambiguous identifiers

Ambiguity arises when a view uses the same name as the underlying type, which is fairly common.
```
WITH
    User := User {
        name,
        foo := 'bar'
    }
SELECT User {...User};

# or worse, the view has no explicitly defined shape
WITH
    User := (SELECT User FILTER .name >= 'B')
SELECT User {...User};
```

We may want to disambiguate whether we're splatting a type or a view, the difference being that splatting types includes *every* property and link of that type, while splatting a view only includes the ones which are immediately defined on the view as a shape annotation.

Perhaps something like:
```
SELECT User {...TYPE User};
SELECT User {...VIEW UserView};

# or maybe 

SELECT User {...User};  # the shortest version includes everything
SELECT User {...VIEW UserView};
```

## 3. Type expressions

Also, if an object type is what's allowed in a splat, then type expressions should be allowed there as well:
```
SELECT User {
    ...Named,
    ...Activatable  # has a property active -> bool
};

# should be nearly equivalent to the above, 
# except for order of properties in serialization
SELECT User {
    ...Named & Activatable  
};
```

## 4. Order of elements

What order do splatted elements appear in? Do we guarantee anything about order of shape elements at all? GraphQL, for example, does specify that the output should contain the fields in the same order as they appear in in the input document. It also specifies order for fields defined in fragments, but GraphQL has the advantage of always having fragments right in the input document, so the order is not terribly obscure.

Rather than relying on "definition order" (which may be non-trivial to infer for complex types extending other types and thus difficult for the end-user to actually predict and rely on), we can specify that a *type* splat always enforces alphabetical order of the splatted elements. At least this way, the order is always well defined and predictable. Incidentally this also works for splatting union and intersection types. Splatting a view might stick to the "definition order", though (although this may be an undesirable inconsistency).

**Comment by ctmattice** [comment]
I come from the multi-value no-sql world, my specif flavor is Rocket Softwares' D3 database system. I'm rather new here but in my world:
SELECT USER by itself would return all users and their fields per user.

In your example:
SELECT User {
    ...Named,
    ...Activatable  # has a property active -> bool
};
That would return a complete listing of users and their ...Named and ...Activatable fields only .

and the example:
SELECT User {...VIEW UserView};
could be used to create a procedural call to sort the users by a specific field or fields, by a non-existant key value, or what ever.

It would be nice to keep the specifc order of an item in tack, that way if you decide to create a specific language in edgeql to do more complex functions, (such as mvbasic in my world), it would be easier to read/write/edit a specific record.

**Comment by ctmattice** [comment]
Sorry misspoke above SELECT USERS by itself would only return all record IDs, from that point they could be edited/read/written to etc.

If we did a SELECT USERS RETURNING (OR CAPTURING) then we could return or capture result for a later operation.

**Comment by 1st1** [comment]
Reviving the discussion here.

> 2. Ambiguous identifiers
> Ambiguity arises when a view uses the same name as the underlying type, which is fairly common.

I don't think any of the listed examples warrant complication of the design and adding `...VIEW` or `...TYPE` splats. Just don't shadow your types with query aliases. And if you do, it's a trivial thing to fix. At least I'll be -1 on implementing this feature from the beginning -- this type of syntax should come from real world examples and user requests.

> 3. Type expressions

> ```SELECT User {
>    ...Named & Activatable  
> };```


I don't see why not, but we don't necessarily need that in 1.0. If we need it at all, I'd use parens after ... to improve readability, as in:

```
SELECT User {
    ...(Named & Activatable)
};```

> What order do splatted elements appear in? Do we guarantee anything about order of shape elements at all? 

Should be in the order the fields are described in the query or in the schema. We must preserve order to the best of our ability, and AFAIK we fully control the order.

> Rather than relying on "definition order" (which may be non-trivial to infer for complex types extending other types and thus difficult for the end-user to actually predict and rely on)

I don't fully agree with that. Predicting order in a complicated diamond inheretance is hard, but it's quite obvious for simple cases (and can be documented).  

>  At least this way, the order is always well defined and predictable.

It's not predictable. Altering a parent type with a new property would screw things up in absolutely the same way.  And ordering would make output to look really weird in repl in simple cases. I'm -1 on that.

**Comment by 1st1** [comment]
> I come from the multi-value no-sql world, my specif flavor is Rocket Softwares' D3 database system. I'm rather new here but in my world:

Thanks for the pointer, I'll take a look at D3.

> Sorry misspoke above SELECT USERS by itself would only return all record IDs

Almost the same in EdgeDB -- we'd return "User" object with only one field, its ID.

> It would be nice to keep the specifc order of an item in tack,

Yes, I agree.

**Comment by haikyuu** [comment]
Maybe we could start iterating on the usage and API at the query builder level. We could add an API for selecting all attributes to the ts query builder. 
```
let query = e.select(e.Movie, movie => ({
 "*": true
}}
let query = e.select(e.Movie, movie => ({
 "*": {
  properties: true,
  links: true // would compile to {id: true} for each link
  computed: true
}
}}
```

**Comment by adriangb** [comment]
That sounds like a really good idea @haikyuu . I imagine introducing something like into the query builder would have lower development cost and lower misstep cost than introducing it into the core EdgeQL syntax. Good way to test out the concept.

**Comment by MarcusCemes** [comment]
> That sounds like a really good idea @haikyuu . I imagine introducing something like into the query builder would have lower development cost and lower misstep cost than introducing it into the core EdgeQL syntax. Good way to test out the concept.

Just a suggestion from a new user, I was looking for a way to quickly inspect objects inside of the egdedb interactive REPL and found my way to this issue.

I find it useful during development to be able to poke around the database and play with objects, modify them, delete them, before the application is completely working. People who have tried Elixir and iex will understand how amazing solid introspection can be. At the moment it's a lot of typing to select properties, even if autocomplete is outstanding.

```text
edgedb> select User filter .username = "Alice";
{default::User {id: d5ceabf2-c252-11ec-8820-9bb3eb36643f}}
```

This is not very useful when the cardinality is one, especially if a User has a _lot_ of table properties to keep them localised in memory/on disk.

It would be a lot more ergonomic to either have a wildcard operator or just be a lot more verbose in the REPL, possibly run a count on links, ...

```text
edgedb> select User filter .username = "Alice";
{default::User {
    id: d5ceabf2-c252-11ec-8820-9bb3eb36643f,
    username: 'Alice',
    sessions: <12 default::Session>
}}
```

I also understand the wish to keep EdgeQL as "unpolluted" as possible, perhaps this could be an extension of the REPL, or a GUI browser such as Prisma Studio (even if I find REPL more ergonomic)?

**Comment by haikyuu** [comment]
@MarcusCemes great idea. The studio (in alpha) has a REPL and this would be a great addition there as well

**Comment by colinhacks** [comment]
This has been added to the JavaScript query builder with the following API.

```typescript
e.select(e.Movie, ()=>({
  ...e.Movie['*']
}))
```

This will fetch all properties of `Movie`; links must be specified separately. This API/behavior is subject to change if/when native splat functionality lands in EdgeQL.

**Comment by haikyuu** [comment]
There is also a syntax that seems nice in [duckdb](https://duckdb.org/2022/05/04/friendlier-sql.html)
```
SELECT * EXCLUDE (jar_jar_binks, midichlorians) FROM star_wars
```

And it would be nice to mark fields as excluded from `*` at schema level 
```
required property password -> str { excluded := true };
```

**Comment by msullivan** [comment]
Alright we are hoping to squeeze this into 3.0. Proposal is up here: https://github.com/edgedb/rfcs/pull/77

**Comment by msullivan** [comment]
Implemented in #5080!
